### PR TITLE
`azurerm_servicebus_namespace` - support for the `premium_messaging_partitions` property

### PR DESCRIPTION
**Pull request by xiaxyi**
part of the MS vanguard, the premium namespace is now support messaging partitions, add the feature to terraform as well.

acc tests:
```
--- PASS: TestAccAzureRMServiceBusNamespace_premiumCapacity (107.95s)
--- PASS: TestAccAzureRMServiceBusNamespace_premiumMessagingPartition (107.95s)

```

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/24597

For the accepted value and the validation part, in theory, we shouldn't add 0 into the acceptable list because the feature only applies to servicebus premium namespace, not to the standard/ basic namespace, however, api will add 0 value for the standard/basic namespaces, as a result, we'll see diff like 0->null during the second apply. 

For the default value, 0 is for the standard/basic namespace while 1 is for the premium namespace. we need to add the validation to check if users are trying to set 0 for the premium namespace, if they do, we throw error before the api does. Things are the same for the standard/ basic namespace, but we check the  >0 value.

Product information: https://learn.microsoft.com/en-us/azure/service-bus-messaging/enable-partitions-premium

### CODE REVIEW

File: website/docs/r/servicebus_namespace.html.markdown [hunk 1]
```diff
@@ -51,6 +51,8 @@ The following arguments are supported:
 
 * `capacity` - (Optional) Specifies the capacity. When `sku` is `Premium`, capacity can be `1`, `2`, `4`, `8` or `16`. When `sku` is `Basic` or `Standard`, capacity can be `0` only.
 
+* `premium_messaging_partitions` - (Optional) Specifies the messaging partitions. When `sku` is `Premium`, messaging partitions can be `1`, `2`, `4`. When `sku` is `Basic` or `Standard`, messaging partitions can be `0` only. Changing this forces a new resource to be created.
```
Comment (jackofallops) [inline website/docs/r/servicebus_namespace.html.markdown:54]:
```suggestion
* `premium_messaging_partitions` - (Optional) Specifies the number of Premium Messaging Partitions. Only valid when `sku` is `Premium`. Possible values include `0`, `1`, `2`, and `4`. Defaults to `0`. Changing this forces a new resource to be created.
```

File: internal/services/servicebus/servicebus_namespace_resource_test.go [hunk 2]
```diff
@@ -455,6 +467,28 @@ resource "azurerm_servicebus_namespace" "test" {
...
+  location = "%s"
+}
+
+resource "azurerm_servicebus_namespace" "test" {
+  name                         = "acctestservicebusnamespace-%d"
+  location                     = azurerm_resource_group.test.location
+  resource_group_name          = azurerm_resource_group.test.name
+  sku                          = "Premium"
+  capacity                     = 2
+  premium_messaging_partitions = 0
```
Comment (jackofallops) [inline internal/services/servicebus/servicebus_namespace_resource_test.go:487]: This defaults to `0` so should be omitted here?
```suggestion
```

File: internal/services/servicebus/servicebus_namespace_resource_test.go [hunk 3]
```diff
@@ -139,6 +139,17 @@ func TestAccAzureRMServiceBusNamespace_premiumCapacity(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMServiceBusNamespace_premiumMessagingPartition(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_servicebus_namespace", "test")
+	r := ServiceBusNamespaceResource{}
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config:      r.premiumMessagingPartition(data),
+			ExpectError: regexp.MustCompile("Service Bus SKU \"Premium\" only supports `premium_messaging_partitions` of 1, 2, 4"),
```
Comment (jackofallops) [inline internal/services/servicebus/servicebus_namespace_resource_test.go:148]: This doesn't match the schema validation above?

File: internal/services/servicebus/servicebus_namespace_resource.go [hunk 4]
```diff
@@ -336,6 +344,16 @@ func resourceServiceBusNamespaceCreateUpdate(d *pluginsdk.ResourceData, meta int
 		parameters.Sku.Capacity = utils.Int64(int64(capacity.(int)))
 	}
 
+	if premiumMessagingUnit := d.Get("premium_messaging_partitions"); premiumMessagingUnit != nil {
+		if !strings.EqualFold(sku, string(namespaces.SkuNamePremium)) && premiumMessagingUnit.(int) > 0 {
+			return fmt.Errorf("Premium messaging partition is not supported by service Bus SKU %q and it can only be set to 0", sku)
+		}
+		if strings.EqualFold(sku, string(namespaces.SkuNamePremium)) && premiumMessagingUnit.(int) == 0 {
+			return fmt.Errorf("Service Bus SKU %q only supports `premium_messaging_partitions` of 1, 2, 4", sku)
+		}
```
Comment (jackofallops) [inline internal/services/servicebus/servicebus_namespace_resource.go:353]: This doesn't match the schema validation above? The property defaults to `0` also.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Thanks @xiaxyi - There's some comments to take a look at below, as it appears there is some confusion about allowed values?  Can you take a look?

File: internal/services/servicebus/servicebus_namespace_resource.go (hunk 4 above)
Comment (xiaxyi) [reply internal/services/servicebus/servicebus_namespace_resource.go:353]: This is for the premium namespace, 0 is not a valid value for premium namespace.

File: internal/services/servicebus/servicebus_namespace_resource_test.go (hunk 3 above)
Comment (xiaxyi) [reply internal/services/servicebus/servicebus_namespace_resource_test.go:148]: The validation error is for the premium namespaces

**Comment by xiaxyi** [comment]
Thanks @jackofallops for reviewing this pr!

The feature is only available in service bus premium namespace but we still need to set it to 0 for the standard/ basic namespace and premium namespace that's created in an unsupported region which not included in below regions:
image

For those premium namespace that's created in the feature-available region, the default value is 1, despite of the 0 value we put in the request body, like 
image

**Review by katbyte — CHANGES_REQUESTED** [review]
We have some test fialures

```
------- Stdout: -------
=== RUN   TestAccAzureRMServiceBusNamespacePairing_basic
=== PAUSE TestAccAzureRMServiceBusNamespacePairing_basic
=== CONT  TestAccAzureRMServiceBusNamespacePairing_basic
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: Service Bus SKU "Premium" only supports `premium_messaging_partitions` of 1, 2, 4
        
          with azurerm_servicebus_namespace.primary_namespace_test,
          on terraform_plugin_test.tf line 31, in resource "azurerm_servicebus_namespace" "primary_namespace_test":
          31: resource "azurerm_servicebus_namespace" "primary_namespace_test" {
        
        
        Error: Service Bus SKU "Premium" only supports `premium_messaging_partitions` of 1, 2, 4
        
          with azurerm_servicebus_namespace.secondary_namespace_test,
          on terraform_plugin_test.tf line 39, in resource "azurerm_servicebus_namespace" "secondary_namespace_test":
          39: resource "azurerm_servicebus_namespace" "secondary_namespace_test" {
        
--- FAIL: TestAccAzureRMServiceBusNamespacePairing_basic (48.64s)
FAIL
```

**Review by katbyte — APPROVED** [review]
LGTM ⛈️

**Pull request merged**